### PR TITLE
[query] Allow query to consolidate series by tags

### PR DESCRIPTION
**Pull request by arnikola**
**What this PR does / why we need it**:
Fixes an issue that arose if tag ID schemes were changed leading to duplicate series.

### CODE REVIEW

File: src/x/serialize/decoder.go [hunk 1]
```diff
@@ -237,3 +239,17 @@ func (d *decoder) Duplicate() ident.TagIterator {
...
 	return iter
 }
+
+func (d *decoder) Hash() (uint32, error) {
+	if d.checkedData == nil {
+		return 0, nil
+	}
+	if d.digest == 0 {
+		digest := adler32.New()
+		if _, err := digest.Write(d.checkedData.Bytes()); err != nil {
```
Comment (robskillington) [inline src/x/serialize/decoder.go:249]: Unfortunately won't be able to depend on this being equivalent to other Hash methods on other tag iterators, if we want to depend on equality with another set of tags (that could be a non-encoded set of tags) we need to generate the hash the same way, rolling hash of name+value in order that tags appear).

File: src/cmd/services/m3coordinator/downsample/id_pool_types.go [hunk 2]
```diff
@@ -215,6 +218,25 @@ func (p *rollupIDProvider) Duplicate() ident.TagIterator {
 	return duplicate
 }
 
+func (p *rollupIDProvider) Hash() (uint32, error) {
+	if len(p.tagPairs) == 0 {
+		return 0, nil
+	}
+	if p.digest == 0 {
+		digest := adler32.New()
+		for _, t := range p.tagPairs {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/id_pool_types.go:227]: Instead of looping over tag pairs it might be easier to refactor ".Current()" to be ".current(index int)" and then do a for loop from zero to the end and ask for the tag at each index?

Otherwise we'll be missing the special tags we insert as part of the rollup provider and only be using the base tags for this hash?

File: src/cmd/services/m3coordinator/downsample/id_pool_types.go [hunk 3]
```diff
@@ -215,6 +218,25 @@ func (p *rollupIDProvider) Duplicate() ident.TagIterator {
 	return duplicate
 }
 
+func (p *rollupIDProvider) Hash() (uint32, error) {
+	if len(p.tagPairs) == 0 {
+		return 0, nil
+	}
+	if p.digest == 0 {
+		digest := adler32.New()
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/id_pool_types.go:226]: Instead of allocating an adler32 hash, can we use the zero-alloc variant we have?
https://github.com/m3db/stackadler32#stackadler32

Comment (robskillington) [reply src/cmd/services/m3coordinator/downsample/id_pool_types.go:226]: Here's an example of us using it (we actually use constructor from `src/dbnode/digest` I guess so we can swap later if we want to), you need to make the digest equal the return of the `Update` function:
https://github.com/m3db/m3/blob/master/src/dbnode/client/session.go#L3105:L3118
```go
		var (
			d        = digest.NewDigest()
			expected = uint32(*checksum)
		)
		if merged := block.Segments.Merged; merged != nil {
			d = d.Update(merged.Head).Update(merged.Tail)
		} else {
			for _, s := range block.Segments.Unmerged {
				d = d.Update(s.Head).Update(s.Tail)
			}
		}
		if actual := d.Sum32(); actual != expected {
			return fmt.Errorf("block checksum is bad: expected=%d, actual=%d", expected, actual)
		}
```

File: src/cmd/services/m3coordinator/downsample/tags.go [hunk 4]
```diff
@@ -115,6 +118,29 @@ func (t *tags) Duplicate() ident.TagIterator {
...
+func (t *tags) Hash() (uint32, error) {
+	if t.digest == 0 {
+		l := len(t.values)
+		if r := len(t.names); r != l {
+			return 0, fmt.Errorf("len values %d does not match len names %d", l, r)
+		}
+		if l == 0 {
+			return 0, nil
+		}
+		digest := adler32.New()
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/tags.go:130]: Here too, can we use our zero-alloc `stackadler32` variant?

File: src/dbnode/network/server/tchannelthrift/convert/convert.go [hunk 5]
```diff
@@ -497,9 +500,26 @@ func (w *writeTaggedIter) Duplicate() ident.TagIterator {
 	return &writeTaggedIter{
 		rawRequest: w.rawRequest,
 		currentIdx: -1,
+		digest:     w.digest,
 	}
 }
 
+func (w *writeTaggedIter) Hash() (uint32, error) {
+	if w.digest == 0 {
+		digest := adler32.New()
```
Comment (robskillington) [inline src/dbnode/network/server/tchannelthrift/convert/convert.go:509]: Here too, can we use our zero-alloc `stackadler32` variant?

File: src/dbnode/storage/index/convert/convert.go [hunk 6]
```diff
@@ -408,3 +411,22 @@ func (t *tagIter) Duplicate() ident.TagIterator {
 	}
 	return &dupe
 }
+
+func (t *tagIter) Hash() (uint32, error) {
+	if t.done || len(t.docFields) == 0 {
+		return 0, nil
+	}
+	if t.digest == 0 {
+		digest := adler32.New()
```
Comment (robskillington) [inline src/dbnode/storage/index/convert/convert.go:420]: Here too, can we use our zero-alloc `stackadler32` variant?

File: src/query/storage/m3/consolidators/fetch_result_map.go [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+		},
+		equals: func(x, y encoding.SeriesIterator) bool {
+			// NB: fail fast on errors.
+			err := wrapper.multiErr.FinalError()
+			if err != nil {
+				return false
+			}
+
+			// NB: succeed fast if possible with matching IDs.
+			if x.ID() == y.ID() {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/fetch_result_map.go:97]: Hm, think you want `x.ID().Equal(y)` yeah?

File: src/query/storage/m3/consolidators/fetch_result_map.go [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+	for _, results := range w.resultMap.Iter() {
+		result = append(result, results.value)
+	}
+
+	return result
+}
+
+func (w *fetchResultMapWrapper) get(
+	it encoding.SeriesIterator,
+) (multiResultSeries, bool) {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/fetch_result_map.go:55]: Maybe it's worth making this potentially return an error too since you can run into an error during get(...) set(..)? I know there's an explicit ".err()" but maybe it's best to just always potentially return an error from get/set (then callers don't need to know they need to remember to call ".err()", etc?

File: src/query/storage/m3/consolidators/fetch_result_map.go (hunk 7 above)
Comment (robskillington) [reply src/query/storage/m3/consolidators/fetch_result_map.go:97]: Also, wonder if we want to bother doing this check? It might be expensive to always be doing a bytes.Compare(...) and a tag comparison each time?

File: src/query/storage/m3/consolidators/fetch_result_map.go [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+			if err != nil {
+				addErr("hashing y iterator failed", err)
+				return false
+			}
+
+			if xHash != yHash {
+				return false
+			}
+
+			xTags, yTags = xTags.Duplicate(), yTags.Duplicate()
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/fetch_result_map.go:125]: Yeah so... we probably need to avoid duplication here for comparison. 

I know this is painful and will require us to implement ".Index()" and ".ResetAtIndex(...)" methods on "ident.TagIterator" but I think we will need to do that.

Otherwise we will do a huge allocation here each time we look up in the map. And there is lock contention on the pools as well, so entire goroutine can be paused/de-scheduled each time we need to do a map get/set possibly.

I may already have this change somewhere in an old PR, let me have a look.

File: src/query/storage/m3/consolidators/id_dedupe_map.go [hunk 10]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/query/storage"
+)
+
+type idDedupeMap struct {
+	fanout QueryFanoutType
+	series map[string]multiResultSeries
+}
+
+func newIDDedupeMap(fanout QueryFanoutType, size int) fetchDedupeMap {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/id_dedupe_map.go:35]: Could you possibly use the `fetchResultMapWrapper` for ID de-duping too and just change the hash and equals FN?

Comment (robskillington) [reply src/query/storage/m3/consolidators/id_dedupe_map.go:35]: That would avoid all the string allocation this current de-dupe method incurs.

File: src/query/storage/m3/consolidators/multi_iterator.go [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+
+	currTs := it.current.dp.Timestamp
+	for {
+		if len(it.indexData) == 0 {
+			return false
+		}
+
+		// NB: take off value corresponding to next ordered iterator.
+		first := it.indexData[0]
+		copy(it.indexData, it.indexData[1:])
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_iterator.go:125]: Interesting, was it not worth while seeing if we could use the already stack based "iterators.go"?

It has the whole concept of "values" and "earliest" and applies the different timestamp de-duping policies. 

Is there a blocker for using it as the parent de-duplicater?
https://github.com/m3db/m3/blob/master/src/dbnode/encoding/iterators.go

File: src/x/serialize/decoder.go [hunk 12]
```diff
@@ -237,3 +239,17 @@ func (d *decoder) Duplicate() ident.TagIterator {
 	}
 	return iter
 }
+
+func (d *decoder) Hash() (uint32, error) {
+	if d.checkedData == nil {
+		return 0, nil
+	}
+	if d.digest == 0 {
+		digest := adler32.New()
```
Comment (robskillington) [inline src/x/serialize/decoder.go:248]: Here too, can we use our zero-alloc `stackadler32` variant?

File: src/query/storage/m3/consolidators/fetch_result_map.go (hunk 7 above)
Comment (arnikola) [reply src/query/storage/m3/consolidators/fetch_result_map.go:97]: The bytes.Compare is much cheaper than reading all the tags, so better to short circuit early if IDs match (which is the case almost all of the time)

File: src/cmd/services/m3coordinator/downsample/id_pool_types.go [hunk 13]
```diff
@@ -215,6 +218,25 @@ func (p *rollupIDProvider) Duplicate() ident.TagIterator {
 	return duplicate
 }
 
+func (p *rollupIDProvider) Hash() (uint32, error) {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/id_pool_types.go:221]: Should we remove this now longer being used for consolidation?

File: src/cmd/services/m3coordinator/downsample/flush_handler.go [hunk 14]
```diff
@@ -178,8 +179,8 @@ func (w *downsamplerFlushHandlerWriter) Write(
 				Value:     mp.Value,
 			}},
 			Unit: convert.UnitForM3DB(mp.StoragePolicy.Resolution().Precision),
-			Attributes: storage.Attributes{
-				MetricsType: storage.AggregatedMetricsType,
+			Attributes: consolidators.Attributes{
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/flush_handler.go:182]: Hm, can we keep these as "storage" attributes? Is it possible to put any types that depend on these types next to each other (if it was a circular dependency problem)?

File: src/cmd/services/m3coordinator/downsample/tags.go [hunk 15]
```diff
@@ -115,6 +118,29 @@ func (t *tags) Duplicate() ident.TagIterator {
 	return &tags{idx: -1, names: t.names, values: t.values}
 }
 
+func (t *tags) Hash() (uint32, error) {
```
Comment (robskillington) [inline src/cmd/services/m3coordinator/downsample/tags.go:121]: Should we remove this now longer being used for consolidation?

File: src/dbnode/client/session_write_tagged_test.go [hunk 16]
```diff
@@ -618,3 +618,4 @@ func (e *erroredTagIter) Close()                       {}
 func (e *erroredTagIter) Len() int                     { return 0 }
 func (e *erroredTagIter) Remaining() int               { return 0 }
 func (e *erroredTagIter) Duplicate() ident.TagIterator { return e }
+func (e *erroredTagIter) Hash() (uint32, error)        { return 0, nil }
```
Comment (robskillington) [inline src/dbnode/client/session_write_tagged_test.go:621]: Should we remove this now longer being used for consolidation?

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+func (it *seriesIteratorAccumulator) ID() ident.ID {
+	return it.id
+}
+
+func (it *seriesIteratorAccumulator) Namespace() ident.ID {
+	return it.nsID
+}
+
+func (it *seriesIteratorAccumulator) Tags() ident.TagIterator {
+	it.err = errors.New("cannot get tags from series iterator accumulator")
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:100]: Maybe we should just avoid exposing "Tags()" on the "SeriesIteratorAccumulator" interface at all?

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+		if err := iter.Err(); err != nil {
+			it.err = err
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (it *seriesIteratorAccumulator) Replicas() []MultiReaderIterator {
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:144]: Same here maybe we just remove the "Replicas()" call from "SeriesIteratorAccumulator"?

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 19]
```diff
@@ -0,0 +1,256 @@
...
+
+func (it *seriesIteratorAccumulator) Reset(opts SeriesIteratorOptions) {
+	if len(it.seriesIterators) < 1 {
+		it.err = errors.New("cannot reset on empty series accumulator")
+		return
+	}
+
+	it.id = opts.ID
+	it.nsID = opts.Namespace
+	for _, it := range it.seriesIterators[1:] {
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:177]: Hm, why only close > 1?  Also should this check that each entry is not nil? (since in "Close()" they are nil'd out but slice is not shrunk `it.seriesIterators[idx] = nil`)

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 20]
```diff
@@ -0,0 +1,256 @@
...
+		it.Close()
+	}
+
+	it.iters.reset()
+	first := it.seriesIterators[0]
+	first.Reset(opts)
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:183]: Is this guaranteed to not be nil?

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 17 above)
Comment (robskillington) [reply src/dbnode/encoding/series_iterator_accumulator.go:100]: If these methods are just going to return errors, I think it might be better to just declare a smaller interface that we can use and just update all places that currently reference a SeriesIterator with a SeriesIteratorAccmulator?

File: src/dbnode/encoding/series_iterator_accumulator_test.go [hunk 21]
```diff
@@ -0,0 +1,164 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator_test.go:1]: nit 2020

File: src/query/storage/m3/consolidators/id_dedupe_map.go [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+
+package consolidators
+
+import (
+	"fmt"
+
+	"github.com/m3db/m3/src/dbnode/encoding"
+)
+
+type idDedupeMap struct {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/id_dedupe_map.go:29]: Do we need this anymore?

File: src/query/storage/m3/consolidators/match_type.go [hunk 23]
```diff
@@ -0,0 +1,71 @@
...
+
+package consolidators
+
+import (
+	"fmt"
+	"strings"
+)
+
+const (
+	defaultMatchType MatchType = MatchIDs
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/match_type.go:29]: If we do some microbenchmarks and it's faster doing MatchTags, maybe we should default to that? Not sure what will be faster, bytes + string alloc and map[string]value or just bytes alloc + genny map.

File: src/query/storage/m3/consolidators/types.go [hunk 24]
```diff
@@ -0,0 +1,174 @@
+// Copyright (c) 2018 Uber Technologies, Inc.
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/types.go:1]: nit 2020

File: src/query/storage/m3/consolidators/multi_iterator.go [hunk 25]
```diff
@@ -0,0 +1,285 @@
...
+	d[idx] = insertData
+
+	d = append(d[:idx], insertData)
+	d = append(d, d[:idx]...)
+}
+
+// NB: enforce multiIterator implements encoding.SeriesIterator.
+var _ encoding.SeriesIterator = (*multiIterator)(nil)
+
+type multiIterator struct {
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_iterator.go:103]: Don't need this anymore yeah?

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 19 above)
Comment (arnikola) [reply src/dbnode/encoding/series_iterator_accumulator.go:205]: So this would close every iterator except the first, which is instead reset, and set to be the `head` of the accumulator

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 20 above)
Comment (arnikola) [reply src/dbnode/encoding/series_iterator_accumulator.go:211]: `first` doesn't get closed, and the accumulator necessarily has at least 1 iterator, so this should be safe

File: src/dbnode/encoding/ostream.go [hunk 26]
```diff
@@ -251,12 +240,7 @@ func (os *ostream) Reset(buffer checked.Bytes) {
 	}
 }
 
-// Rawbytes returns the Osteam's raw bytes. Note that this does not transfer ownership
-// of the data and bypasses the checked.Bytes accounting so callers should:
-//     1. Only use the returned slice as a "read-only" snapshot of the data in a context
-//        where the caller has at least a read lock on the ostream itself.
-//     2. Use this function with care.
```
Comment (robskillington) [inline src/dbnode/encoding/ostream.go:258]: Hm, can we move these docs somewhere else? They're fairly valuable.

Perhaps to the interface itself?

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 27]
```diff
@@ -0,0 +1,281 @@
...
+	err             error
+	firstNext       bool
+	closed          bool
+}
+
+// SeriesAccumulatorOptions are options for a SeriesIteratorAccumulator.
+type SeriesAccumulatorOptions struct {
+	// RetainTags determines if tags should be preserved after the accumulator is
+	// exhausted. If set to true, the accumulator retains a copy of the tags.
+	RetainTags bool
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:52]: Could we use Rewind for this case?

Just for helping me review the PR, when do we need to use RetainTags?

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 19 above)
Comment (robskillington) [reply src/dbnode/encoding/series_iterator_accumulator.go:200]: I guess I don't quite understand `Reset()`.

Is this meant to be called before or after "Add(....)"?

Usually "Reset(...)" means we pulled it from the pool and we're about to use it..

File: src/dbnode/encoding/ostream.go (hunk 26 above)
Comment (arnikola) [reply src/dbnode/encoding/ostream.go:258]: These should be moved to [encoding/types](https://github.com/m3db/m3/pull/2370/files/4af783dd2a56cd7075b3db9c9ecb7d1946aea0b1#diff-2e6541aee73116ef7c93a9618f5063c4R406) to standardize with our other comment style; I can move it back if preferred though

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 27 above)
Comment (arnikola) [reply src/dbnode/encoding/series_iterator_accumulator.go:52]: So it's a bit weird; since `Accumulator` uses the base `encoding.interface`, as part of its iteration, it closes exhausted sub-iterators which frees their tag iterators, too. Functionally, this means that if you call `Tags()` on the Accumulator after iterating through the values, you wouldn't be able to get any Tags back, since it wouldn't have any more data.

File: src/dbnode/encoding/series_iterator_accumulator.go [hunk 28]
```diff
@@ -0,0 +1,281 @@
...
+func (it *seriesIteratorAccumulator) Replicas() ([]MultiReaderIterator, error) {
+	if l := len(it.seriesIterators); l != 1 {
+		return nil, fmt.Errorf("cannot get replicas for accumulated series "+
+			"iterators: need 1 iterator, have %d", l)
+	}
+
+	return it.seriesIterators[0].Replicas()
+}
+
+func (it *seriesIteratorAccumulator) Reset(opts SeriesIteratorOptions) {
```
Comment (robskillington) [inline src/dbnode/encoding/series_iterator_accumulator.go:192]: After complications talked about seems like maybe just it.err = not supported is best route for now.

File: src/query/storage/m3/consolidators/multi_fetch_result.go [hunk 29]
```diff
@@ -114,48 +136,70 @@ func (r *multiResult) FinalResult() (SeriesFetchResult, error) {
 	r.Lock()
 	defer r.Unlock()
 
-	result := SeriesFetchResult{Metadata: r.metadata}
+	// result := SeriesFetchResult{Metadata: r.metadata}
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_fetch_result.go:139]: Can remove?

File: src/dbnode/encoding/ostream.go (hunk 26 above)
Comment (robskillington) [reply src/dbnode/encoding/ostream.go:258]: Great, no that sounds good.

File: src/query/storage/m3/consolidators/multi_fetch_result.go [hunk 30]
```diff
@@ -114,48 +136,70 @@ func (r *multiResult) FinalResult() (SeriesFetchResult, error) {
...
+		// If incoming list somehow shorter, shrink stored list.
+		r.mergedTags = r.mergedTags[:lenNext]
+	}
 
-	i := 0
-	for _, res := range r.dedupeMap {
+	for i, res := range dedupedList {
 		r.mergedIterators.SetAt(i, res.iter)
-		i++
+		r.mergedTags[i] = res.tags
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_fetch_result.go:178]: Do we actually use `r.mergedTags` anywhere? I see it constructed but I only ever see us returning r.mergedIterators and r.metadata?

File: src/query/storage/m3/consolidators/multi_fetch_tags_result.go [hunk 31]
```diff
@@ -35,10 +35,11 @@ type multiSearchResult struct {
 	meta      block.ResultMetadata
 	err       xerrors.MultiError
 	seenIters []client.TaggedIDsIterator // track known iterators to avoid leaking
+	// FIXME: take a series iter version instead.
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_fetch_tags_result.go:38]: Do we need to address this before this PR is merged?

File: src/query/storage/m3/storage.go [hunk 32]
```diff
@@ -327,12 +310,17 @@ func (s *m3storage) fetchCompressed(
 				)
 			}
 
+			if err != nil {
+				result.ReportError(err)
```
Comment (robskillington) [inline src/query/storage/m3/storage.go:314]: Should the below not happen if err != nil?

File: src/query/storage/m3/consolidators/multi_fetch_tags_result.go (hunk 31 above)
Comment (arnikola) [reply src/query/storage/m3/consolidators/multi_fetch_tags_result.go:38]: Ah wait this was inaccurate, since this stuff works on tags only, it's immune to mismatch ID issues

File: src/dbnode/encoding/series_iterator_accumulator.go (hunk 28 above)
Comment (arnikola) [reply src/dbnode/encoding/series_iterator_accumulator.go:192]: Sounds good; can look into putting this back later if the need arises 👍

File: src/query/storage/m3/consolidators/multi_fetch_result.go (hunk 30 above)
Comment (arnikola) [reply src/query/storage/m3/consolidators/multi_fetch_result.go:178]: Yeah, looks like it was a bug, would cause tags to be recalculated each time
nice catch!

File: src/query/storage/m3/storage.go (hunk 32 above)
Comment (arnikola) [reply src/query/storage/m3/storage.go:314]: Nice catch

File: scripts/docker-integration-tests/query_fanout/m3coordinator-cluster-a.yml [hunk 33]
```diff
@@ -58,3 +58,7 @@ carbon:
 
 tagOptions:
   idScheme: quoted
+
+query:
+  consolidation:
+    matchType: tags
```
Comment (robskillington) [inline scripts/docker-integration-tests/query_fanout/m3coordinator-cluster-a.yml:64]: nit: Maybe add a comment to why it's valuable to test this configuration with query fanout?

File: src/query/storage/m3/consolidators/multi_fetch_result.go [hunk 34]
```diff
@@ -86,25 +102,30 @@ func (r *multiResult) Close() error {
...
+				attrs = append(attrs, res.attrs)
+			}
 		}
 	} else {
-		i := 0
-		for _, res := range r.dedupeMap {
-			attrs[i] = res.attrs
-			i++
-		}
+		attrs = []storagemetadata.Attributes{}
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_fetch_result.go:128]: Is the else clause necessary? 

Seems like leaving it as just nil and zero length instead of `attrs = []storagemetadata.Attributes{}` is just fine?

File: src/query/storage/m3/consolidators/multi_fetch_result.go [hunk 35]
```diff
@@ -114,48 +135,74 @@ func (r *multiResult) FinalResult() (SeriesFetchResult, error) {
...
-	for _, res := range r.dedupeMap {
+	for i, res := range dedupedList {
 		r.mergedIterators.SetAt(i, res.iter)
-		i++
+		r.mergedTags[i] = res.tags
+	}
+
+	pTags := make([]*models.Tags, 0, len(r.mergedTags))
+	for _, t := range r.mergedTags {
+		pTags = append(pTags, &t)
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/multi_fetch_result.go:181]: This will take a ref to the loop var =[

Perhaps we add a sanity test where length is greater than zero of results coming back at the `multi_fetch_result.go` level of this code path so this would be caught?

Comment (robskillington) [reply src/query/storage/m3/consolidators/multi_fetch_result.go:181]: Repro here:
https://play.golang.org/p/8TrDqZ__AX4

File: src/query/storage/m3/storage.go [hunk 36]
```diff
@@ -327,17 +311,22 @@ func (s *m3storage) fetchCompressed(
...
+				return
```
Comment (robskillington) [inline src/query/storage/m3/storage.go:316]: Now that this is possible to receive an error, before returning at the bottom of top level function `fetchCompressed` here should we do:
```go
if err := result.Err(); err != nil {
  return nil, err
}
return result, nil
```

Seems more sane than letting someone run into the `iter.Err` later perhaps?

File: src/query/storage/m3/consolidators/tag_dedupe_map_test.go [hunk 37]
```diff
@@ -0,0 +1,143 @@
...
+
+	"github.com/golang/mock/gomock"
+	"github.com/m3db/m3/src/dbnode/encoding"
+	"github.com/m3db/m3/src/dbnode/ts"
+	"github.com/m3db/m3/src/query/models"
+	"github.com/m3db/m3/src/query/storage/m3/storagemetadata"
+	"github.com/m3db/m3/src/x/ident"
+	xtest "github.com/m3db/m3/src/x/test"
+	xtime "github.com/m3db/m3/src/x/time"
+	"github.com/stretchr/testify/assert"
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/tag_dedupe_map_test.go:35]: nit: Separate imports?

File: src/query/storage/m3/consolidators/tag_dedupe_map_test.go [hunk 38]
```diff
@@ -0,0 +1,143 @@
...
+	"github.com/stretchr/testify/require"
+)
+
+func varifyDedupeMap(
```
Comment (robskillington) [inline src/query/storage/m3/consolidators/tag_dedupe_map_test.go:39]: nit: verify?

File: src/x/serialize/decoder.go [hunk 39]
```diff
@@ -237,3 +237,8 @@ func (d *decoder) Duplicate() ident.TagIterator {
 	}
 	return iter
 }
+
+func (d *decoder) Rewind() {
+	d.checkedData.IncRef()
```
Comment (robskillington) [inline src/x/serialize/decoder.go:242]: Think you want to DecRef after "d.Reset(...)" otherwise this we will never reach NumRef() == 0.

File: src/x/serialize/decoder_test.go [hunk 40]
```diff
@@ -127,6 +129,39 @@ func TestDecodeSimple(t *testing.T) {
...
+		require.False(t, d.Next())
+		require.NoError(t, d.Err())
+		copy(exBytes[i*len(printedTags):], printedTags)
+		d.Rewind()
+	}
+
+	assert.Equal(t, exBytes, acBytes)
+	assert.Equal(t, string(exBytes), string(acBytes))
+
+	d.Close()
```
Comment (robskillington) [inline src/x/serialize/decoder_test.go:162]: Want to get the checked bytes and check that `b.NumRef()` is zero here?

**Comment by BertHartm** [comment]
Would this help #2184?
